linera-core/src/chain_worker/state/mod.rs

```diff
@@ -352,30 +327,51 @@ where
         Ok(())
     }
 
-    /// Returns the blobs requested by their `blob_ids` that are in the chain manager's pending blobs
-    /// and checks that they are otherwise in storage.
-    async fn get_blobs_and_checks_storage(
+    /// Returns the blobs required by the given executed block. The ones that are not passed in
+    /// are read from the chain manager or from storage.
+    async fn get_required_blobs(
         &self,
-        blob_ids: &HashSet<BlobId>,
-    ) -> Result<Vec<Blob>, WorkerError> {
-        let pending_blobs = &self.chain.manager.get().pending_blobs;
-
-        let mut found_blobs = Vec::new();
-        let mut missing_blob_ids = Vec::new();
-        for blob_id in blob_ids {
-            if let Some(blob) = pending_blobs.get(blob_id) {
-                found_blobs.push(blob.clone());
-            } else {
-                missing_blob_ids.push(*blob_id);
+        executed_block: &ExecutedBlock,
+        blobs: &[Blob],
```

Comment:
The comment says
> The ones that are not passed in are read from the chain manager or from storage.

Is it correct to say then that `blobs` are _published_ blobs?

Reply:
No, it can't be 🤔 the `executed_block.required_blob_ids()` returns blobs published in that block.

Reply:
This method has no opinion on that: It just tries to use the blobs that were already passed in first, before looking them up in the chain manager or storage.

In some call sites they happen to be the published blobs, yes, because that's what's included in a block proposal.

Reply:
What's confusing to me is how often we call methods like `handle_*` with and w/o blobs. What's the rule? For example - when the callstack starts with `handle_lite_certificate` we always pass empty blob array (even into the `handle_confirmed_certificate` method eventually) but if the same stack starts with `fully_handle_certificate` then it _might_ contain some blobs.

This is, in my opinion, too convoluted. We probably should slice the code so that it's easier to get "what's what".

Reply:
There's no _strict_ rule: _Conceptually_, we have to send the certificate and all relevant blobs (those that are published and those that are read for the first time on the chain).

However, in many cases we can expect the validator to already have some of that data, so as an optimization, we send them parts of it and see if they complain. If they do, we send everything.

E.g. if a validator is one of the signatories, we expect them to already have that certificate in the cache.

For the read blobs I think we often just send nothing and then retry; we should probably be smarter about that, but that's for another PR. I want to make things correct first, then optimize (or improve the optimization that's there).